### PR TITLE
backupccl: move datadriven framework to its own file

### PR DESCRIPTION
**Pull request by adityamaru**
This change moves the datadriven framework and its utility
methods to its own file.

Additionally, it adds more instructions to the datadriven driver, and ports a few of the tests in `backupcc/backup_test.go` to datadriven tests.

Informs: #77129

Release note: None

Release justification: non-production code changes

### CODE REVIEW

**Comment by adityamaru** [comment]
Apologies for the large diff but I think the easiest way to review this will be to see the final state of `datadriven_test.go`, in particular the header comment to get an idea of the new instructions.

Each commit then has the old tests in `backup_test.go`, and their new ported tests in the testdata files.

The last commit is just a reorganization of the testdata files to split them out into different test drivers.

**Comment by adityamaru** [comment]
friendly ping!

File: pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors [hunk 1]
```diff
@@ -10,9 +10,12 @@ SET use_declarative_schema_changer = 'off';
 SET CLUSTER SETTING jobs.debug.pausepoints = 'schemachanger.before.exec';
 CREATE DATABASE d;
 CREATE TABLE d.foo (id INT);
+----
```
Comment (msbutler) [inline pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors:13]: did the diff for this data driven test accidentally creep into the first commit?

Comment (adityamaru) [reply pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors:13]: yeah I think I got excited and added a `expect-pausepoint` subclause to `exec-sql`. This doesn't live in `exec-sql` in the final state though so I can revert this diff.

**Review by msbutler — COMMENTED** [review]
I feel like this PR should get split up. 
- A quick PR for moving the data driven driver to its own file. After doing this, it would be easier for me to migrate show tests to the data driven framework. Perhaps this PR also contains a few of the new bells and whistles you want to add to the driver.
- A PR for each test refactor-- I think each test migration requires its own close unique examination, each with a potentially different set of reviewers, no?

File: pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors (hunk 1 above)
Comment (adityamaru) [reply pkg/ccl/backupccl/testdata/backup-restore/backup-dropped-descriptors:13]: fwiw in a future PR I'm going to make `exec-sql` default to `require.NoError()` and if a statement expects an error they will have to add an additional `exec-sql expects-error`. This is more canonical with the way all other datadriven drivers work today.

**Comment by msbutler** [comment]
Right, that does make sense! The new features of the driver should get bundled in the same PR(s) as the test they apply to.

**Comment by adityamaru** [comment]
So basically a PR per commit then?

**Pull request closed without merging**